### PR TITLE
Not found state

### PR DESCRIPTION
**Pull request by jakubpycha**
If stateasync do not load state then return 404.

### CODE REVIEW

**Comment by richardschneider** [comment]
Hi @jakubpycha and thanks for taking time to improve this project!

I'm unclear about what you are trying to fix.  It would help if you added a test to show the issue.

Could the problem be
> the conditions that check an `etag` should first verify that `stateasycn` has returned a?n etag

**Comment by jakubpycha** [comment]
Hi @richardschneider ,
The problem is when you try to make an update over a record that does not exist (returns 404). Then the default getResourceStateAsync returns an empty etag, then process check is failed and the request remains unsolved.

**Comment by richardschneider** [comment]
Here's a test I wrote (in test/resourceState) to prove the problem

```js
    it('should 404 when resource does not exist', done => {
        request(server)
            .put("/something-else")
            .set('authorization', 'Basic YWxpY2U6eHl6enk=')
            .set('if-match', resourceEtag)
            .expect(404)
            .end(done);
    });
```

The test just times out.

File: lib/index.js [hunk 1]
```diff
@@ -29,11 +36,24 @@ function process(req, res, next, options) {
...
             if (status != 200)
                 return options.error(status, 'A requested precondition failed', req, res);
             next();
+        })
+        .catch(err => {
+            if (err.status) {
+                options.error(err.status, err.message, req, res);
+            } else {
+                options.error(500, err.message, res, res);
+            }
```
Comment (richardschneider) [inline lib/index.js:56]: Yes the `catch` is needed; good find.  Just a nitpick, I prefer
```
options.error(err,status ? 500, err.message, req, res)
```

File: lib/index.js [hunk 2]
```diff
@@ -29,11 +36,24 @@ function process(req, res, next, options) {
...
     options
         .stateAsync(req)
-        .then(state => check(req, state, preconditions))
+        .then(state => {
+            if (state.etag) {
+                return check(req, state, preconditions)
+            } else {
+                throw new ProcessError(404, 'Not found');
+            }
+        })
```
Comment (richardschneider) [inline lib/index.js:45]: As the failed tests indicate, we can't do etag check here.  

Need to change `ifMatch` and `ifNotMatch (in preconditons.js) to return 404 if etag is undefined.

File: lib/index.js [hunk 3]
```diff
@@ -2,6 +2,13 @@
 
 const allPreconditions = require('./preconditions');
 
+class ProcessError extends Error {
```
Comment (richardschneider) [inline lib/index.js:5]: Don't need this, see following comments.

**Review by richardschneider — CHANGES_REQUESTED** [review]
- catching a promise rejection is required, well done.
- checking for missing etag must be done in the preconditions.js

**Pull request closed without merging**